src/librustdoc/html/render/sorted_template.rs

```diff
@@ -0,0 +1,139 @@
+use std::collections::BTreeSet;
+use std::fmt;
+use std::marker::PhantomData;
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+/// Append-only templates for sorted, deduplicated lists of items.
+///
+/// Last line of the rendered output is a comment encoding the next insertion point.
+#[derive(Debug, Clone)]
+pub(crate) struct SortedTemplate<F> {
```

Comment:
I think the file format is logically part of the template, because the template either contains html or js content. Like if you create a js template it's always going to contain js content, and I think the type system should enforce that